suggest use of `kube explain <resource>` in kube get output

juanvallejo opened the pull request:
**Release note**:

``` release-note
release-note-none
```

This patch improves usability flow, making it easier for a user to
discover the command `kube explain <resource>` through `kube get` output.
##### After

```
$ kube get
You must specify the type of resource to get. Valid resource types include:
   * componentstatuses (aka 'cs')
   * configmaps (aka 'cm')
   * daemonsets (aka 'ds')
   * deployments (aka 'deploy')
   * events (aka 'ev')
   * endpoints (aka 'ep')
   * horizontalpodautoscalers (aka 'hpa')
   * ingress (aka 'ing')
   * jobs
   * limitranges (aka 'limits')
   * nodes (aka 'no')
   * namespaces (aka 'ns')
   * petsets (alpha feature, may be unstable)
   * pods (aka 'po')
   * persistentvolumes (aka 'pv')
   * persistentvolumeclaims (aka 'pvc')
   * quota
   * resourcequotas (aka 'quota')
   * replicasets (aka 'rs')
   * replicationcontrollers (aka 'rc')
   * secrets
   * serviceaccounts (aka 'sa')
   * services (aka 'svc')
error: Required resource not specified.
Use "kubectl explain <resource>" for a detailed description of that resource (e.g. kubectl explain pods).
See 'kubectl get -h' for help and examples.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31818)

<!-- Reviewable:end -->

janetkuo commented:
ok to test

juanvallejo committed:
suggest use of `kube explain <resource>` in kube get output

This patch improves usability flow, making it easier for a user to
discover the command `kube explain <resource>` through `kube get`
output.

janetkuo commented:
@kubernetes/kubectl 

Suggest not hardcoding "kubectl". You should be able to get the command name from `cmd.Parent().CommandPath()` (cobra command library)

AdoHe commented:
@juanvallejo please do not use `kubectl` directly.

juanvallejo commented:
@janetkuo @AdoHe Thanks for the feedback! I replaced `kubectl` with `cmd.Root().CommandPath()` https://github.com/kubernetes/kubernetes/pull/31818/files#diff-f23cc6c004fbd8a713f8465465142ab8R258

janetkuo commented on pkg/kubectl/cmd/get.go:
```diff
@@ -168,7 +180,8 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
 	if len(args) == 0 && cmdutil.IsFilenameEmpty(options.Filenames) {
 		fmt.Fprint(errOut, "You must specify the type of resource to get. ", valid_resources)
-		return cmdutil.UsageError(cmd, "Required resource not specified.")
+		usageString := fmt.Sprintf("Required resource not specified.\nUse \"%s explain <resource>\" for a detailed description of that resource (e.g. %[1]s explain pods).", options.FullCmdName)
```
You don't need to pass `FullCmdName` in. You can use the passed-in `cmd` and do the `cmd. cmds.Root().CommandPath()` here.

juanvallejo replied:
Done! PTAL https://github.com/kubernetes/kubernetes/pull/31818/commits/fea3756aeeaa1c6ca9b0611c2ef868807e506ecf

janetkuo commented on pkg/kubectl/cmd/get.go:
```diff
@@ -168,7 +168,14 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
-		return cmdutil.UsageError(cmd, "Required resource not specified.")
+
+		fullCmdName := cmd.Parent().CommandPath()
+		if fullCmdName == "" {
```
If `fullCmdName == ""` we shouldn't assume that the user can use `kubectl explain`. 

We also need a check to see if `<parent_command> explain` exists, something like:
```go
func isExplainExist(cmd *cobra.Command) bool {
		for _, c := range cmd.Parent().Commands() {
			if c.Name() == "explain" {
				return true
			}
		}
		return false 
}
```

and we only print the suggestion when `len(fullCmdName)>0` (i.e. `fullCmdName != ""`) and the explain command is found:
```go
usageString := "Required resource not specified."
if len(fullCmdName)>0 && isExplainExist(cmd) {
	usageString = fmt.Sprintf("%s\nUse ...", usageString, fullCmdName)
}
```

janetkuo commented:
One last suggestion. 
@juanvallejo would you open another PR (or a separate commit in this PR, whichever you prefer) to change the similar thing in `kubectl exec` to what this PR does? (i.e. remove passing "kubectl" / command path to `NewCmdExec` and do the similar check)

juanvallejo commented:
@janetkuo Thanks for the feedback! 
> would you open another PR (or a separate commit in this PR, whichever you prefer) to change the similar thing in kubectl exec to what this PR does?

Sure, I'll add this in a new commit

juanvallejo committed:
add check to NewCmdExec before printing suggestion

checks that the "describe" command and a parent command path exist
before printing suggestion to use the describe command to list
containers in a pod.

janetkuo commented on pkg/kubectl/cmd/exec.go:
```diff
@@ -265,7 +267,11 @@ func (p *ExecOptions) Run() error {
+			if len(p.SuggestedCmdUsage) > 0 {
+				usageString = fmt.Sprintf("%s\n%s", usageString, p.SuggestedCmdUsage)
+			}
+			fmt.Fprintf(p.Err, "%s", usageString)
```
nit: add `\n` at the end

janetkuo commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -716,3 +716,15 @@ func ObjectListToVersionedObject(objects []runtime.Object, version unversioned.G
+
+// IsCommandExists receives a pointer to a cobra command and a target string.
+// Returns true if the target string is found in the list of sibling commands.
+func IsCommandExists(cmd *cobra.Command, targetCmdName string) bool {
```
nit: on second thought we should name it `IsSibCommandExists` to make it more precise (update the name in the comment as well)

janetkuo commented:
Only nits, lgtm otherwise. Feel free to squash the 1st, 2nd and 3rd commits into one.

janetkuo commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -716,3 +716,15 @@ func ObjectListToVersionedObject(objects []runtime.Object, version unversioned.G
+
+// IsSubCommandExists receives a pointer to a cobra command and a target string.
+// Returns true if the target string is found in the list of sibling commands.
+func IsSubCommandExists(cmd *cobra.Command, targetCmdName string) bool {
```
typo: `IsSibCommandExists`, sib, not sub

janetkuo replied:
or should we use sibling instead of sib? In case people confuse sib with sub

juanvallejo replied:
Ah, I would prefer `sibling` instead, it would be a bit easier to understand

janetkuo commented:
LGTM, thanks much!

janetkuo commented:
@k8s-bot gke e2e test this

janetkuo commented:
@k8s-bot gci gce e2e test this
